pkg/codegen/hcl2/README.md

```diff
@@ -0,0 +1,205 @@
+  * A *block* of *type* `output` with a list of *labels* comprising the string
+    `"o"`, and a *body* containing the following children:
+    * An *attribute* named `value` with the *expression* `r.name`.
+
+At the syntactic level, a PCL program is no more than a piece of HCL syntax.
+Notions of "resources", "outputs", and so on, do not exist. In order to, for
+instance, assign *semantics* to a piece of PCL syntax (e.g. "the expression
+`r.name` refers to the `name` attribute of the `r` resource"), or to decide
+whether or not some PCL program is in some sense "well-formed", we need to
+[*bind*](pcl-binding) the program to produce a *semantic model*. The part of
+binding that can be performed at the level of the syntax tree itself (that is,
+without understanding higher-level concepts such as resources or outputs) is
+known as [*type checking*](pcl-type-checking).
+
+(pcl-type-checking)=
+## Type checking
+
+The `pkg/codegen/hcl2/model` package exports types and functions for *type
+checking* parsed HCL syntax to produce a semantic *model*. Specifically, it
+provides functionality for type checking *expressions*, since these are the only
+parts of the syntax that can be reasoned about without coupling the syntax to
+higher-level concepts such as resources and outputs. Type checking thus
+comprises a part of the wider [binding](pcl-binding) process, whereby a program
+that is being bound is type checked, with the type checker receiving as input a
+*scope* that helps it reason about the types of e.g. references it encounters in
+expressions. In this way, a reference can be understood to e.g. have an object
+type without having to couple the type checker to how a resource's type is an
+object consisting of its output properties.
+
+This package's types mirror (and typically wrap) the types of the AST, adding
+semantic information such as types and resolved references. So for instance, a
+`model.LiteralValueExpression`, which corresponds to a literal value such as the
+string `"foo"` or the number `42`, contains:
+
+* a reference to the `hclsyntax.LiteralValueExpr` underpinning it in the syntax
+  tree;
+* an evaluated `cty.Value` representing the value of the literal; and
+* a `model.Type` representing the type of the literal.
+
+(pcl-type-system)=
+### Type system
+
+This section covers the aforementioned extensions PCL's type system makes to the
+[*HCL syntax-agnostic information
+model*](https://github.com/hashicorp/hcl/blob/v2.3.0/spec.md).
+
+#### Types
+
+##### Primitive types
+
+PCL adds the *int* primitive type. An *int* is an arbitrary-precision integer
+value. Implementations *must* make full-precision values available to consumer
+applications for interpretation into any suitable integer representation.
+Implementations may in practice implement *int*s with limited precision so long
+as the following constraints are met:
+
+* Integers are represented with at least 256 bits.
```

Comment:
Where is this requirement coming from?  I don't think we're using 256bit integers in Go anywhere, or am I missing something?

Reply:
I just moved this from the old `spec.md`. I presume it means that if you don't have arbitrary precision integers, you must have at least 256 bits. Go has `math/big` for arbitrary numbers, which we appear to use in `pkg/codegen/hcl2`, so I _assume_ we are sidestepping this requirement here.